### PR TITLE
Container surrogate for UseEmulator for Azure Storage.

### PR DESCRIPTION
**Pull request by mitchdenny**
Container surrogate for Azure Storage. The idea is that with this optional callback you can manipuate the underlying resource as if it were a container when using the emulator. This was a model proposed at the App Model offsite.

The builder in the callback on `UseEmulator` is for type `AzureStorageEmulatorContainerResource` which derives from `ContainerResource` to activate all of those extension methods. The annotations a proxied through to the underlying resource.

Here is an example of using the code to enable persistent storage:

```csharp
var builder = DistributedApplication.CreateBuilder(args);

var storage = builder.AddAzureStorage("storage").UseEmulator(container =>
{
    container.UsePersistence();
});

var blobs = storage.AddBlobs("blobs");

builder.AddProject<Projects.AzureStorageEndToEnd_ApiService>("api")
       .WithReference(blobs);

builder.Build().Run();
```

### CODE REVIEW

File: playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+var builder = DistributedApplication.CreateBuilder(args);
+
+var storage = builder.AddAzureStorage("storage").UseEmulator(configureContainer: builder =>
+{
+    // Temporary, we will have an extension method that allows setting persistence and specifying a local
+    // directory (for example).
+    builder.WithVolumeMount(Path.Combine(Environment.CurrentDirectory, "data"), "/data", VolumeMountType.Bind, false);
```
Comment (davidfowl) [inline playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs:10]: We should make volumes relative to the apphost directory.

File: playground/AzureStorageEndToEnd/AzureStorageEndToEnd.ApiService/Program.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Azure.Storage.Blobs;
+
+var builder = WebApplication.CreateBuilder(args);
+
+int insertionRows = builder.Configuration.GetValue<int>("InsertionRows", 1);
```
Comment (davidfowl) [inline playground/AzureStorageEndToEnd/AzureStorageEndToEnd.ApiService/Program.cs:8]:
```suggestion
int insertionRows = builder.Configuration.GetValue("InsertionRows", 1);
```

Comment (mitchdenny) [reply playground/AzureStorageEndToEnd/AzureStorageEndToEnd.ApiService/Program.cs:8]: Actually I have to remove this. Copypasta.

File: playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs (hunk 1 above)
Comment (mitchdenny) [reply playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs]: Done.

File: src/Aspire.Hosting/DistributedApplicationResourceBuilder.cs [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
 namespace Aspire.Hosting;
 
-internal sealed class DistributedApplicationResourceBuilder<T>(IDistributedApplicationBuilder applicationBuilder, T resource) : IResourceBuilder<T> where T : IResource
+public sealed class DistributedApplicationResourceBuilder<T>(IDistributedApplicationBuilder applicationBuilder, T resource) : IResourceBuilder<T> where T : IResource
```
Comment (davidfowl) [inline src/Aspire.Hosting/DistributedApplicationResourceBuilder.cs:8]: Nice! Should we make it easier to make a builder from a resource? Extension method?

Comment (mitchdenny) [reply src/Aspire.Hosting/DistributedApplicationResourceBuilder.cs:8]: Yes. I was thinking the same thing.

Comment (mitchdenny) [reply src/Aspire.Hosting/DistributedApplicationResourceBuilder.cs]: Done.

File: playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+var builder = DistributedApplication.CreateBuilder(args);
+
+var storage = builder.AddAzureStorage("storage").UseEmulator(configureContainer: builder =>
+{
+    builder.UsePersistence("c:\\code\\mydata");
```
Comment (davidfowl) [inline playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs:8]: Undo?

File: playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+var builder = DistributedApplication.CreateBuilder(args);
+
+var storage = builder.AddAzureStorage("storage").UseEmulator(configureContainer: builder =>
```
Comment (davidfowl) [inline playground/AzureStorageEndToEnd/AzureStorageEndToEnd.AppHost/Program.cs:6]:
```suggestion
var storage = builder.AddAzureStorage("storage").UseEmulator(container  =>
```

File: src/Aspire.Hosting.Azure/AzureResourceExtensions.cs [hunk 6]
```diff
@@ -160,26 +161,39 @@ private static void WriteQueueStorageToManifest(ManifestPublishingContext contex
...
+    /// <summary>
+    /// Enables persistence in the Azure Storage emulator.
+    /// </summary>
+    /// <param name="builder">The builder for the <see cref="AzureStorageEmulatorResourceContainerSurrogate"/>.</param>
+    /// <param name="path">Relative path to the AppHost where emulator storage is persisted between runs.</param>
+    /// <returns>A builder for the <see cref="AzureStorageEmulatorResourceContainerSurrogate"/>.</returns>
+    public static IResourceBuilder<AzureStorageEmulatorResourceContainerSurrogate> UsePersistence(this IResourceBuilder<AzureStorageEmulatorResourceContainerSurrogate> builder, string? path = null)
+    {
+        path = path ?? $".azurite/{builder.Resource.Name}";
+        var fullyQualifiedPath = Path.GetFullPath(path);
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/AzureResourceExtensions.cs:192]: Make it relative to builder.ApplicationBuilder.AppHostDirectory

File: src/Aspire.Hosting.Azure/AzureStorageEmulatorContainerResource.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+
+namespace Aspire.Hosting.Azure;
+
+public class AzureStorageEmulatorResourceContainerSurrogate(AzureStorageResource innerResource) : ContainerResource(innerResource.Name), IResource
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/AzureStorageEmulatorContainerResource.cs:8]:
```suggestion
public class AzureStorageEmulatorResource(AzureStorageResource innerResource) : ContainerResource(innerResource.Name), IResource
```

File: src/Aspire.Hosting.Azure/AzureResourceExtensions.cs [hunk 8]
```diff
@@ -157,29 +158,39 @@ private static void WriteQueueStorageToManifest(ManifestPublishingContext contex
...
+    /// <summary>
+    /// Enables persistence in the Azure Storage emulator.
+    /// </summary>
+    /// <param name="builder">The builder for the <see cref="AzureStorageEmulatorResource"/>.</param>
+    /// <param name="path">Relative path to the AppHost where emulator storage is persisted between runs.</param>
+    /// <returns>A builder for the <see cref="AzureStorageEmulatorResource"/>.</returns>
+    public static IResourceBuilder<AzureStorageEmulatorResource> UsePersistence(this IResourceBuilder<AzureStorageEmulatorResource> builder, string? path = null)
+    {
+        path = path ?? $".azurite/{builder.Resource.Name}";
+        var fullyQualifiedPath = Path.GetFullPath(path, builder.ApplicationBuilder.AppHostDirectory);
```
Comment (davidfowl) [inline src/Aspire.Hosting.Azure/AzureResourceExtensions.cs:189]: Wow TIL. I was doing `Path.GetFullPath(Path.Combine(base, path))`

**Review by davidfowl — APPROVED** [review]
Very nice

File: src/Aspire.Hosting.Azure/AzureResourceExtensions.cs (hunk 8 above)
Comment (mitchdenny) [reply src/Aspire.Hosting.Azure/AzureResourceExtensions.cs:189]: I only just noticed it too.

File: src/Aspire.Hosting.Azure/AzureStorageEmulatorResource.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Aspire.Hosting.ApplicationModel;
+
+namespace Aspire.Hosting.Azure;
+
+public class AzureStorageEmulatorResource(AzureStorageResource innerResource) : ContainerResource(innerResource.Name), IResource
```
Comment (IEvangelist) [inline src/Aspire.Hosting.Azure/AzureStorageEmulatorResource.cs:8]: All public APIs should have triple slash comments. We should not merge any PRs without them, perhaps we can fail the build or something if they're missing? //cc @davidfowl and @mitchdenny

Comment (davidfowl) [reply src/Aspire.Hosting.Azure/AzureStorageEmulatorResource.cs:8]: Yes, we should start to do that.

Comment (davidfowl) [reply src/Aspire.Hosting.Azure/AzureStorageEmulatorResource.cs:8]: cc @joperezr

**Pull request merged**